### PR TITLE
introduce incubator/drone

### PR DESCRIPTION
**Pull request by christian-roggia**
This is a continuation of the PRs https://github.com/kubernetes/charts/pull/821 and https://github.com/kubernetes/charts/pull/353.

TODO:

- [x] upgrade to version 0.8.1
- [x] standardize source to helm 2.7
- [x] test service.type=ClusterIP
- [x] test service.type=NodePort
- [x] test service.type=LoadBalancer
- [x] test ingress
- [x] test bare install
- [x] test source control provider installation
- [x] README

The Chart has been tested on a cluster configured via kubeadm with niginx-ingress installed and configured. The cluster version is 1.8.2. This cluster was installed on bare metal (Ubuntu 16.04) and unfortunately I do not have access to a LoadBalancer right now.

If you can contribute to the Chart by testing it on your cluster with different configurations and report if everything is working correctly it would be really appreciated.

Also, I really need help with the README, if anyone can improve it please let me know.

### CODE REVIEW

File: incubator/drone/templates/deployment-agent.yaml [hunk 1]
```diff
@@ -0,0 +1,52 @@
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: {{ template "drone.fullname" . }}-agent
+  labels:
+    app: {{ template "drone.fullname" . }}-agent
+    chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+    release: "{{ .Release.Name }}"
+    heritage: "{{ .Release.Service }}"
```
Comment (unguiculus) [inline incubator/drone/templates/deployment-agent.yaml]: Add component label: `component: agent`

File: incubator/drone/templates/deployment-agent.yaml [hunk 2]
```diff
@@ -0,0 +1,52 @@
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: {{ template "drone.fullname" . }}-agent
+  labels:
+    app: {{ template "drone.fullname" . }}-agent
```
Comment (unguiculus) [inline incubator/drone/templates/deployment-agent.yaml]: Should be: `app: {{ template "drone.name" . }}`

File: incubator/drone/templates/deployment-agent.yaml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+    release: "{{ .Release.Name }}"
+    heritage: "{{ .Release.Service }}"
+spec:
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        app: {{ template "drone.fullname" . }}-agent
```
Comment (unguiculus) [inline incubator/drone/templates/deployment-agent.yaml]: Should be: `app: {{ template "drone.name" . }}`

File: incubator/drone/templates/deployment-agent.yaml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+        release: "{{ .Release.Name }}"
+        heritage: "{{ .Release.Service }}"
```
Comment (unguiculus) [inline incubator/drone/templates/deployment-agent.yaml]: Only `app`, `release`, and `component` labels should go here.

File: incubator/drone/templates/deployment-server.yaml [hunk 5]
```diff
@@ -0,0 +1,54 @@
+{{- if hasKey .Values.server.env "DRONE_PROVIDER" }}
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: {{ template "drone.fullname" . }}-server
+  labels:
+    app: {{ template "drone.fullname" . }}-server
```
Comment (unguiculus) [inline incubator/drone/templates/deployment-server.yaml]: Should be: `app: {{ template "drone.name" . }}`

File: incubator/drone/templates/deployment-server.yaml [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+    release: "{{ .Release.Name }}"
+    heritage: "{{ .Release.Service }}"
```
Comment (unguiculus) [inline incubator/drone/templates/deployment-server.yaml]: Add component label: `component: server`

File: incubator/drone/templates/deployment-server.yaml [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+spec:
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        app: {{ template "drone.fullname" . }}-server
+        chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+        release: "{{ .Release.Name }}"
+        heritage: "{{ .Release.Service }}"
```
Comment (unguiculus) [inline incubator/drone/templates/deployment-server.yaml]: Only `app`, `release`, and `component` labels should go here.

File: incubator/drone/templates/secrets.yaml [hunk 8]
```diff
@@ -0,0 +1,16 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  name: {{ template "drone.fullname" . }}
+  labels:
+    app: {{ template "drone.fullname" . }}
```
Comment (unguiculus) [inline incubator/drone/templates/secrets.yaml]: `app: {{ template "drone.name" . }}`

File: incubator/drone/templates/service.yaml [hunk 9]
```diff
@@ -0,0 +1,24 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ template "drone.fullname" . }}
+  labels:
+    app: {{ template "drone.fullname" . }}
```
Comment (unguiculus) [inline incubator/drone/templates/service.yaml]: `app: {{ template "drone.name" . }}`

File: incubator/drone/templates/service.yaml [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+  - name: http
+    port: {{ .Values.service.httpPort }}
+    targetPort: 8000
+{{- if hasKey .Values.service "nodePort" }}
+    nodePort: {{ .Values.service.nodePort }}
+{{- end }}
+  - name: grpc
+    port: 9000
+    targetPort: 9000
+  selector:
```
Comment (unguiculus) [inline incubator/drone/templates/service.yaml:22]:
```yaml
app: {{ template "drone.name" . }}
release: {{ .Release.Name | quote }}
component: server
```

File: incubator/drone/values.yaml [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+  #    memory: 32Mi
+  #    cpu: 40m
+  #  limits:
+  #    memory: 2Gi
+  #    cpu: 1
+
+## Uncomment this if you want to set a specific shared secret between
+## the agents and servers, otherwise this will be auto-generated.
+##
+# shared_secret: supersecret
```
Comment (unguiculus) [inline incubator/drone/values.yaml]: Use camel case: `sharedSecret`

**Review by unguiculus — CHANGES_REQUESTED** [review]

**Review by unguiculus — APPROVED** [review]

**Comment by unguiculus** [comment]
I tested with a LoadBalancer service and Github. I get redirected to Github and then get "cannot register unguiculus. registration closed". Maybe I misconfigured something. I guess this shouldn't stop us from merging?

**Comment by christian-roggia** [comment]
You have to uncomment/set the environment variable `server.env.DRONE_ADMIN="unguiculus"`

**Comment by unguiculus** [comment]
Must have missed that. Working now.

**Comment by steffenmllr** [comment]
@christian-roggia I tried the chart on my fresh `v1.8.3` cluster. Syncing the git Repos worked fine. The build was also triggered. It seems like the communication between the worker doesn't work as expected:

### UI Log
```
+ git init
Initialized empty Git repository in /drone/src/github.com/my/repo/.git/
+ git remote add origin https://github.com/my/repo.git
+ git fetch --no-tags origin +refs/heads/mybranch:
fatal: unable to access 'https://github.com/my/repo.git/': LibreSSL SSL_connect: SSL_ERROR_SYSCALL in connection to github.com:443 
exit status 128
```

## drone-drone-agent log
```
{"time":"2017-11-22T15:18:27Z","level":"debug","message":"request next execution"}
INFO: 2017/11/22 15:18:27 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing dial tcp 10.3.197.142:9000: getsockopt: connection refused"; Reconnecting to {drone-drone:9000 <nil>}
INFO: 2017/11/22 15:18:28 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing dial tcp 10.3.197.142:9000: getsockopt: connection refused"; Reconnecting to {drone-drone:9000 <nil>}
{"time":"2017-11-22T15:19:41Z","level":"debug","repo":"my/repo","build":"1","id":"1","message":"received execution"}
{"time":"2017-11-22T15:19:41Z","level":"debug","repo":"my/repo","build":"1","id":"1","message":"listen for cancel signal"}
{"time":"2017-11-22T15:19:42Z","level":"debug","repo":"my/repo","build":"1","id":"1","image":"plugins/git:latest","stage":"clone","exit_code":0,"exited":false,"message":"update step status"}
{"time":"2017-11-22T15:19:42Z","level":"debug","repo":"my/repo","build":"1","id":"1","image":"plugins/git:latest","stage":"clone","exit_code":0,"exited":false,"message":"update step status complete"}
{"time":"2017-11-22T15:19:47Z","level":"debug","repo":"my/repo","build":"1","id":"1","image":"plugins/git:latest","stage":"clone","message":"log stream opened"}
{"time":"2017-11-22T15:20:41Z","level":"debug","repo":"my/repo","build":"1","id":"1","message":"pipeline lease renewed"}
{"time":"2017-11-22T15:21:41Z","level":"debug","repo":"my/repo","build":"1","id":"1","message":"pipeline lease renewed"}
{"time":"2017-11-22T15:22:41Z","level":"debug","repo":"my/repo","build":"1","id":"1","message":"pipeline lease renewed"}
{"time":"2017-11-22T15:22:51Z","level":"debug","repo":"my/repo","build":"1","id":"1","image":"plugins/git:latest","stage":"clone","message":"log stream copied"}
{"time":"2017-11-22T15:22:51Z","level":"debug","repo":"my/repo","build":"1","id":"1","image":"plugins/git:latest","stage":"clone","message":"log stream uploading"}
{"time":"2017-11-22T15:22:51Z","level":"debug","repo":"my/repo","build":"1","id":"1","image":"plugins/git:latest","stage":"clone","message":"log stream upload complete"}
{"time":"2017-11-22T15:22:51Z","level":"debug","repo":"my/repo","build":"1","id":"1","image":"plugins/git:latest","stage":"clone","message":"log stream closed"}
{"time":"2017-11-22T15:22:51Z","level":"debug","repo":"my/repo","build":"1","id":"1","image":"plugins/git:latest","stage":"clone","exit_code":1,"exited":true,"message":"update step status"}
{"time":"2017-11-22T15:22:51Z","level":"debug","repo":"my/repo","build":"1","id":"1","image":"plugins/git:latest","stage":"clone","exit_code":1,"exited":true,"message":"update step status complete"}
{"time":"2017-11-22T15:22:51Z","level":"debug","repo":"my/repo","build":"1","id":"1","error":"","exit_code":1,"message":"pipeline complete"}
{"time":"2017-11-22T15:22:51Z","level":"debug","repo":"my/repo","build":"1","id":"1","message":"uploading logs"}
{"time":"2017-11-22T15:22:51Z","level":"debug","repo":"my/repo","build":"1","id":"1","message":"uploading logs complete"}
{"time":"2017-11-22T15:22:51Z","level":"debug","repo":"my/repo","build":"1","id":"1","error":"","exit_code":1,"message":"updating pipeline status"}
{"time":"2017-11-22T15:22:51Z","level":"debug","repo":"my/repo","build":"1","id":"1","message":"stop listening for cancel signal"}
{"time":"2017-11-22T15:22:52Z","level":"debug","repo":"my/repo","build":"1","id":"1","message":"updating pipeline status complete"}
{"time":"2017-11-22T15:22:52Z","level":"debug","message":"request next execution"}
{"time":"2017-11-22T15:22:52Z","level":"debug","repo":"my/repo","build":"1","id":"1","message":"pipeline done"}
```


## drone-drone-dind log
```
time="2017-11-22T15:18:24.897826532Z" level=warning msg="could not change group /var/run/docker.sock to docker: group docker not found" 
time="2017-11-22T15:18:24.898018137Z" level=warning msg="[!] DON'T BIND ON ANY IP ADDRESS WITHOUT setting --tlsverify IF YOU DON'T KNOW WHAT YOU'RE DOING [!]" 
time="2017-11-22T15:18:24.908200825Z" level=info msg="libcontainerd: new containerd process, pid: 19" 
time="2017-11-22T15:18:25.995581049Z" level=info msg="Graph migration to content-addressability took 0.00 seconds" 
time="2017-11-22T15:18:25.996235868Z" level=info msg="Loading containers: start." 
time="2017-11-22T15:18:25.997241996Z" level=warning msg="Running modprobe bridge br_netfilter failed with message: modprobe: can't change directory to '/lib/modules': No such file or directory\n, error: exit status 1" 
time="2017-11-22T15:18:25.997910115Z" level=warning msg="Running modprobe nf_nat failed with message: `modprobe: can't change directory to '/lib/modules': No such file or directory`, error: exit status 1" 
time="2017-11-22T15:18:25.998602435Z" level=warning msg="Running modprobe xt_conntrack failed with message: `modprobe: can't change directory to '/lib/modules': No such file or directory`, error: exit status 1" 
time="2017-11-22T15:18:26.194142059Z" level=info msg="Default bridge (docker0) is assigned with an IP address 172.17.0.0/16. Daemon option --bip can be used to set a preferred IP address" 
time="2017-11-22T15:18:26.249548225Z" level=info msg="Loading containers: done." 
time="2017-11-22T15:18:26.255982607Z" level=info msg="Daemon has completed initialization" 
time="2017-11-22T15:18:26.256009207Z" level=info msg="Docker daemon" commit=89658be graphdriver=overlay version=17.05.0-ce 
time="2017-11-22T15:18:26.263192510Z" level=info msg="API listen on [::]:2375" 
time="2017-11-22T15:18:26.263266912Z" level=info msg="API listen on /var/run/docker.sock" 
time="2017-11-22T15:19:47Z" level=warning msg="Running modprobe nf_nat failed with message: `modprobe: can't change directory to '/lib/modules': No such file or directory`, error: exit status 1" 
time="2017-11-22T15:19:47Z" level=warning msg="Running modprobe xt_conntrack failed with message: `modprobe: can't change directory to '/lib/modules': No such file or directory`, error: exit status 1" 
time="2017-11-22T15:22:51.354480591Z" level=error msg="Error closing logger: invalid argument" 
time="2017-11-22T15:22:51.808390049Z" level=error msg="Handler for POST /v1.26/containers/0_8410094534320841776_clone/kill returned error: Cannot kill container 0_8410094534320841776_clone: Container 6ee4d4c94ee7c36a769aeb69e59cad782579dca1837061581f39150fcd71fc6f is not running" 
```

**Comment by christian-roggia** [comment]
Hi @steffenmllr thank you for your report, I installed the drone chart in my cluster and the build was successful. I was therefore unsuccessful in reproducing your bug, my logs are clean and the build did not fail.

Can you open a new issue and provide more information? That would help a lot!

**Comment by steffenmllr** [comment]
Hey @christian-roggia thanks for your reply. I've put everything I know in #2837 - if you need more information let me know. For now I've switched back to my normal k8n deployment of an older drone version which works. Thanks for your work.

**Pull request merged**